src/crypter.rs

```diff
@@ -0,0 +1,145 @@
+use aes_gcm::{
+    aead::{AeadCore, AeadInPlace, KeyInit, OsRng},
+    Aes128Gcm as AesGcm, Error, KeySizeUser, Nonce,
+};
+use base64::{engine::general_purpose::STANDARD, DecodeError, Engine};
+use std::{
+    collections::VecDeque,
+    fmt::{self, Debug, Formatter},
+    iter,
+    str::FromStr,
+    sync::Arc,
+};
+use typenum::marker_traits::Unsigned;
+
+pub type Key = aes_gcm::Key<AesGcm>;
+
+#[derive(Clone)]
+pub struct Crypter(Arc<CrypterInner>);
+
+#[derive(thiserror::Error, Debug, Clone)]
+pub enum CrypterParseError {
+    #[error(transparent)]
+    Base64(#[from] DecodeError),
+
+    #[error("incorrect key length, must be {} bytes after base64 decode", <AesGcm as KeySizeUser>::KeySize::USIZE)]
+    IncorrectLength,
+
+    #[error("at least one key needed")]
+    Missing,
+}
+
+impl FromStr for Crypter {
+    type Err = CrypterParseError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut keys = s
+            .split(',')
+            .map(|s| {
+                STANDARD
+                    .decode(s)
+                    .map_err(CrypterParseError::Base64)
+                    .and_then(|v| Key::from_exact_iter(v).ok_or(CrypterParseError::IncorrectLength))
+            })
+            .collect::<Result<VecDeque<Key>, _>>()?;
+        let current_key = keys.pop_front().ok_or(CrypterParseError::Missing)?;
+        Ok(Self::new(current_key, keys))
+    }
+}
+
+impl Debug for Crypter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Crypter")
+            .field("current_key", &"..")
+            .field("past_keys", &self.0.past_keys.len())
+            .finish()
+    }
+}
+
+#[derive(Clone, Debug)]
+struct CrypterInner {
+    current_key: Key,
+    past_keys: Vec<Key>,
+}
+
+impl From<Key> for Crypter {
+    fn from(key: Key) -> Self {
+        Self::new(key, [])
+    }
+}
+
+impl Crypter {
+    pub fn new(current_key: Key, past_keys: impl IntoIterator<Item = Key>) -> Self {
+        Self(Arc::new(CrypterInner {
+            current_key,
+            past_keys: past_keys.into_iter().collect(),
+        }))
+    }
+
+    pub fn random() -> Self {
+        Self::generate_key().into()
+    }
+
+    pub fn generate_key() -> Key {
+        AesGcm::generate_key(OsRng)
+    }
+
+    pub fn encrypt(&self, associated_data: &[u8], plaintext: &[u8]) -> Result<Vec<u8>, Error> {
+        self.0.encrypt(associated_data, plaintext)
+    }
+
+    pub fn decrypt(
+        &self,
+        associated_data: &[u8],
+        nonce_and_ciphertext: &[u8],
+    ) -> Result<Vec<u8>, Error> {
+        self.0.decrypt(associated_data, nonce_and_ciphertext)
+    }
+}
+
+impl CrypterInner {
+    fn encrypt(&self, associated_data: &[u8], plaintext: &[u8]) -> Result<Vec<u8>, Error> {
+        let nonce = AesGcm::generate_nonce(&mut OsRng);
+        let mut buffer = plaintext.to_vec();
+        AesGcm::new(&self.current_key).encrypt_in_place(&nonce, associated_data, &mut buffer)?;
```

Comment:
nit: it seems that every use of the keys is to pass to `AesGcm::new`. Perhaps `CrypterInner` should instead store these `AesGcm` values to avoid reconstructing them with every operation?

Reply:
I'd love help in figuring out if that's safe to do. I couldn't tell if the AesGcm keeps state between calls that would be cryptographically unsafe to reuse. It does have a "ghash" that it optionally zeroizes, so I assumed that's sensitive data

Reply:
`AeadInPlace::encrypt_in_place()` does not take a mutable reference to the cipher, so it should be fine to reuse the same object. Drilling down, the `ghash` holds an authentication key. (It can also hold the universal hash output it computes, but this is only updated in an ephemeral clone of `self.ghash`.) This is sensitive, but we are also holding the key it's derived from in memory for the lifetime of the program. We should reuse the `AesGcm` object, as it'll be worthwhile to only expand the AES key schedule once.

Reply:
addressed in 2a40071fb31f6d98bef09fe22b3fecc403f36025, we now store AesGcms instead of Keys